feat(algebra/free_algebra): Define a grading

eric-wieser opened the pull request:
The grading takes the form `free_algebra R X →ₐ[R] add_monoid_algebra (free_algebra R X) ℕ`

It's likely that there are more generic lemmas about grading that can be shared with `tensor_algebra` and friends, but that's left to follow-up work

---
<!-- put comments you want to keep out of the PR commit here -->

- [x] depends on #4315 (first commit)
- [x] depends on #4365
- [x] depends on #4603
- [x] depends on #4789

eric-wieser committed:
feat(data/monoid_algebra): Add missing has_coe_to_fun instance

eric-wieser committed:
feat(algebra/free_algebra): Attempt to define a grading

The grading takes the form `free_algebra R X ≃ₐ[R] add_monoid_algebra (free_algebra R X) ℕ`

eric-wieser committed:
Death throes of a failed attempt

eric-wieser committed:
feat(algebra/free_algebra): A few lemmas about grades of free algebras

The grading takes the form `free_algebra R X →ₐ[R] add_monoid_algebra (free_algebra R X) ℕ`

It's likely that there are more generic lemmas about grading that can be shared with `tensor_algebra` and friends, but that's left to follow-up work

eric-wieser commented on src/algebra/free_algebra.lean:
```diff
@@ -295,4 +295,43 @@ begin
 (by { ext, simp, })
 
+
+open_locale big_operators
```
```suggestion
```
Will fix once the dependent PR is in.

eric-wieser commented on src/algebra/free_algebra.lean:
```diff
@@ -295,4 +295,43 @@ begin
+  simp [finsupp.sum_single_index],
+end
+
+noncomputable
+instance : has_coe (add_monoid_algebra (free_algebra R X) ℕ) (free_algebra R X) := ⟨
+  (add_monoid_algebra.sum_id R : add_monoid_algebra (free_algebra R X) ℕ →ₐ[R] (free_algebra R X))
+⟩
+
+@[simp, norm_cast]
+lemma coe_def (x : add_monoid_algebra (free_algebra R X) ℕ) : (x : free_algebra R X) = add_monoid_algebra.sum_id R x := rfl
```
This seemed harmless, and it makes the mouthful that is `add_monoid_algebra.sum_id` not really matter

kbuzzard replied:
You mean the instance? Why not make it a local instance?

eric-wieser commented on src/algebra/free_algebra.lean:
```diff
@@ -295,4 +295,43 @@ begin
+/-- An element of `R` lifted with `algebra_map` has a single grade 0 element -/
+lemma grades_algebra_map (r : R) :
+  grades (algebra_map R (free_algebra R X) r) = finsupp.single 0 (algebra_map R (free_algebra R X) r) :=
+by {unfold grades, simp}
```
I don't know how to avoid the `unfold` here

eric-wieser committed:
feat(algebra/free_algebra): Add an inductive principle

rwbarton commented on src/data/monoid_algebra.lean:
```diff
@@ -679,6 +684,42 @@ lemma alg_hom_ext_iff {R : Type u₃} [comm_semiring k] [add_monoid G]
 ⟨λ h, alg_hom_ext h, by rintro rfl _; refl⟩
 
+/--
+The `alg_hom` which maps from a grading of an algebra `A` back to that algebra.
```
I wouldn't use the word "grading" in this context. Rather, I would call this map the augmentation map of the monoid algebra: https://ncatlab.org/nlab/show/augmentation+ideal#for_group_algebras

For the Lean identifier, either `augmentation` or something like your `sum_id` or perhaps `sum_coeffs` seems fine.

eric-wieser replied:
Would you mind suggesting a complete docstring? My mathematical background is severely lacking.

rwbarton commented on src/data/monoid_algebra.lean:
```diff
@@ -679,6 +684,42 @@ lemma alg_hom_ext_iff {R : Type u₃} [comm_semiring k] [add_monoid G]
+The `alg_hom` which maps from a grading of an algebra `A` back to that algebra.
+
+The proofs here look an awful lot like `add_monoid_algebra.lift`, but that signature does not allow
+us to distinguish `k` and `A`.
```
The general question here is, when A and R are k-algebras, how to describe Hom_{k-Alg}(A[G], R)?
The case A = k is handled by `lift` but in general, when R is noncommutative, the answer isn't so simple. The reason that this `sum_id` map exists is because it's built from the map from G to R = A sending everything to 1, which lands in the center of A.
Noncommutative polynomial evaluation is another special case and I'm not sure we have a totally satisfactory treatment of this special case yet.

eric-wieser replied:
I started trying to generalize this (after cleaning it up in #4338) and I think ran into the difficulty with commutativity you describe. I think I found that I could replace `b • F a` with `F₁ b * F₂ a` provided I have a hypothesis that `∀ a b, F₁ b * F₂ a = F₂ a * F₁ b. That's at least general enough to recover both `lift` and `sum_id`.

eric-wieser replied:
Here's what that looks like:
```
variables (k) [comm_semiring k]
variables (G) [monoid G]
variables (A : Type u₃) [semiring A] [algebra k A]
variables (B : Type*) [semiring B] [algebra k B]

def lift_aux (F : G →* B) (H : A →ₐ[k] B) (FH_comm : ∀ b a, H b * F a = F a * H b) : monoid_algebra A G →ₐ[k] B :=
{
  to_fun := λ f, f.sum (λ a b, H b * F a),
  map_one' := by {
    rw [one_def, sum_single_index],
    { rw [F.map_one, H.map_one, one_mul], },
    { rw [F.map_one, H.map_zero, zero_mul], },
  },
  map_mul' := λ f g,
    begin
      rw [mul_def, finsupp.sum_mul, finsupp.sum_sum_index],
      work_on_goal 1 { intros, rw [H.map_zero, zero_mul], },
      work_on_goal 1 { intros, rw [H.map_add, add_mul], },
      refine finset.sum_congr rfl (λ a ha, _),
      simp only,
      rw [finsupp.mul_sum, finsupp.sum_sum_index],
      work_on_goal 1 { intros, rw [H.map_zero, zero_mul], },
      work_on_goal 1 { intros, rw [H.map_add, add_mul], },
      refine finset.sum_congr rfl (λ a' ha', _),
      simp only,
      rw [sum_single_index, H.map_mul, F.map_mul],
      work_on_goal 1 { intros, rw [H.map_zero, zero_mul], },
      calc H (f a) * H (g a') * (F a * F a')
         = H (f a) * (H (g a') * F a) * F a' : by simp only [mul_assoc]
      ...= H (f a) * (F a * H (g a')) * F a' : by rw FH_comm
      ...= H (f a) * F a * (H (g a') * F a') : by simp only [mul_assoc],
    end,
  map_zero' := sum_zero_index,
  map_add' := λ f g,
    begin
      rw [sum_add_index],
      { intros, rw [H.map_zero, zero_mul], },
      { intros, rw [H.map_add, add_mul], },
    end,
  commutes' := λ r,
    begin
      rw [coe_algebra_map, sum_single_index, F.map_one, mul_one, alg_hom.commutes],
      intros, rw [H.map_zero, zero_mul],
    end, }

/-- Any monoid homomorphism `G →* A` can be lifted to an algebra homomorphism
`monoid_algebra k G →ₐ[k] A`. -/
def lift : (G →* A) ≃ (monoid_algebra k G →ₐ[k] A) :=
{ inv_fun := λ f, (f : monoid_algebra k G →* A).comp (of k G),
  to_fun := λ F, lift_aux k G k A F (algebra.of_id k _) (λ a b, by
    simp [algebra.of_id, algebra.algebra_map_eq_smul_one]),
  left_inv := λ f, begin ext x, simp [lift_aux, sum_single_index] end,
  right_inv := λ F,
    begin
      ext f,
      conv_rhs { rw ← f.sum_single },
      simp [algebra.of_id, algebra.algebra_map_eq_smul_one, lift_aux, ← F.map_smul, finsupp.sum, ← F.map_sum],
    end }
```

rwbarton commented on src/data/monoid_algebra.lean:
```diff
@@ -50,7 +50,7 @@ The monoid algebra over a semiring `k` generated by the monoid `G`.
 endowed with the convolution product.
 -/
-@[derive [inhabited, add_comm_monoid]]
+@[derive [inhabited, add_comm_monoid, has_coe_to_fun]]
```
I'm confused about what's going on here. Judging from lemmas like `mul_single_one_apply`, it already seems to be possible to apply values of `monoid_algebra` type as though they were functions.

eric-wieser replied:
This discussion belongs in #4315

rwbarton commented on src/data/monoid_algebra.lean:
```diff
@@ -225,10 +225,10 @@ instance [ring k] [monoid G] : ring (monoid_algebra k G) :=
 { mul_comm := mul_comm, .. monoid_algebra.ring}
 
-instance [semiring k] : has_scalar k (monoid_algebra k G) :=
+instance {R : Type*} [semiring R] [semiring k] [semimodule R k] : has_scalar R (monoid_algebra k G) :=
```
Can you switch the names `R` and `k` in this instance and the next one? This way around is too confusing.

eric-wieser replied:
I've updated #4365 to use a clearer naming

rwbarton commented on src/data/monoid_algebra.lean:
```diff
@@ -591,10 +596,10 @@ instance [ring k] [add_monoid G] : ring (add_monoid_algebra k G) :=
 { mul_comm := mul_comm, .. add_monoid_algebra.ring}
 
-instance [semiring k] : has_scalar k (add_monoid_algebra k G) :=
+instance {R : Type*} [semiring R] [semiring k] [semimodule R k] : has_scalar R (add_monoid_algebra k G) :=
```
Also switch these

eric-wieser replied:
#4365

eric-wieser committed:
chore(algebra/free_algebra): Use a subtype instead of a struct

eric-wieser committed:
chore(algebra/free_algebra): Disable false positive linter

eric-wieser committed:
chore(algebra/free_algebra): Tidy up coercions

eric-wieser committed:
chore(*): Remove duplication between `sum_id` and `lift`

eric-wieser committed:
refactor(algebra/free_algebra): Replace most of the code with subalgebra

eric-wieser committed:
fix(*): remove squeeze_simp

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into eric-wieser/free_algebra-induction

eric-wieser committed:
chore(*): Tidy the proof

eric-wieser committed:
Add induction for the exterior algebra too

eric-wieser committed:
fix(*): Solve merge mistake

eric-wieser committed:
wip

eric-wieser committed:
chore(algebra/monoid_algebra): Replace `algebra_map'` with `single_(zero|one)_ring_hom`

 `algebra_map'` is now trivially equal to `single_(zero|one)_ring_hom.comp`, so is no longer needed.

eric-wieser committed:
Merge branch 'eric-wieser/single-hom' into eric-wieser/free_algebra-grading

eric-wieser committed:
wip

eric-wieser committed:
wip

eric-wieser committed:
Merge branch 'eric-wieser/free_algebra-induction' into eric-wieser/free_algebra-grading

# Conflicts:
#	src/algebra/free_algebra.lean

eric-wieser committed:
Revert unrelated changes

eric-wieser committed:
Merge branch 'master' of github.com:leanprover-community/mathlib into eric-wieser/free_algebra-grading

eric-wieser committed:
Fix bad merge

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into eric-wieser/free_algebra-grading

eric-wieser committed:
wip

eric-wieser committed:
feat(algebra/algebra/basic): Add `map_finsupp_(sum|prod)` to `alg_(hom|equiv)`

eric-wieser committed:
feat(algebra/algebra/basic): Add `map_finsupp_(sum|prod)` to `alg_(hom|equiv)`

eric-wieser committed:
Merge branch 'eric-wieser/alg_hom.map_finsupp_sum' into eric-wieser/free_algebra-grading

eric-wieser committed:
Cleanup some more

eric-wieser committed:
Merge branch 'eric-wieser/alg_hom.map_finsupp_sum' into eric-wieser/free_algebra-grading

eric-wieser committed:
Fix the proof

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into eric-wieser/free_algebra-grading

eric-wieser commented on src/algebra/monoid_algebra.lean:
```diff
@@ -291,70 +291,90 @@ end algebra
+For a `monoid_algebra A G`, this is an algebra morphism that applies `H` to the `A` part and `F` to
+the `G` part. Multiplication between `H a` and `F g` must be commutative.
+-/
+def lift_aux (F : G →* B) (H : A →ₐ[k] B) (FH_comm : ∀ a g, H a * F g = F g * H a) : monoid_algebra A G →ₐ[k] B :=
```
I suspect this has a lot in common with lemmas like `polynomial.eval₂_ring_hom' - I intend to attempt to unify them in a follow-up PR.

eric-wieser replied:
Looks like @urkud beat me to this refactor

eric-wieser committed:
Cleanup the proofs

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into eric-wieser/free_algebra-grading

eric-wieser commented on src/algebra/monoid_algebra.lean:
```diff
@@ -703,55 +736,73 @@ end algebra
+-/
+def lift_aux (F : multiplicative G →* B) (H : A →ₐ[k] B) (FH_comm : ∀ a g, H a * F g = F g * H a) : add_monoid_algebra A G →ₐ[k] B :=
+{ to_fun := λ f, f.sum (λ g a, H a * F g),
+  map_one' := by {
```
```suggestion   
  -- The proofs here are almost identical to `monoid_algebra.lift_aux`, but use `erw` instead of `rw`
  -- to unfold `multiplicative`
  map_one' := by {
```

eric-wieser committed:
Update src/algebra/monoid_algebra.lean

kbuzzard commented on src/algebra/free_algebra.lean:
```diff
@@ -337,4 +339,88 @@ begin
+-/
+noncomputable
+def grades : (free_algebra R X) →ₐ[R] add_monoid_algebra (free_algebra R X) ℕ :=
+lift R $ finsupp.single 1 ∘ ι R
```
We have some algebra A (the free algebra), and the question is how to usefully state that A is actually an infinite internal direct sum of submodules A_i. This `grades` map here seems to be a map which does the following: if a is in A, let a_i be its i'th "piece". Then a_i is in A_i and all but finitely many a_i are zero. We can regard the a_i as elements of A. Now the a_i is a finite sequence of elements of A. So we can make them into a polynomial a_i X^i. `grades` sends a to this polynomial. I am surprised that this is a useful concept! But one cool thing about it is that it does seem to be a ring homomorphism, rather than just an R-module homomorphism. The next lemma says that if you evaluate the polynomial at 1 then you get back the element you started with. I've been out of the loop for over a month -- sorry if all this conversation has been had already. Is this the way to do grading of an algebra? It still seems a bit weird to me.

eric-wieser replied:
I think my choice not to use `polynomial` was deliberate, since I wasn't really trying to invoke the mental model of one (although admittedly I had not spotted that `polynomial A` is defeq to `add_monoid_algebra A ℕ`).

> I am surprised that this is a useful concept!

It seemed to me like the only (and obvious) way to induce a grading from the universal property alone. I have no idea if this is "how it's done", it's just the way I stumbled upon. I suppose though there's no need for it to be the public interface to the grading, it was just simplest to leave it that way, and the fact it was a ring homomorphism seemed cute, even though there's zero reason to need `grades x * grades y` as an alternate spelling of `grades (x * y)`.

> Is this the way to do grading of an algebra?

I assume you mean for a general algebra?

eric-wieser replied:
> the question is how to usefully state that A is actually an infinite internal direct sum of submodules A_i

I attempted to do this in #4812. The proofs were really quite painful, and I don't think we want to use an API like that until `dfinsupp` is easier to use.

kbuzzard commented on src/algebra/free_algebra.lean:
```diff
@@ -337,4 +339,88 @@ begin
+noncomputable
+instance : has_coe (add_monoid_algebra (free_algebra R X) ℕ) (free_algebra R X) := ⟨
+  (add_monoid_algebra.sum_id R : add_monoid_algebra (free_algebra R X) ℕ →ₐ[R] (free_algebra R X))
+⟩
```
Should this really be an instance? Do we know for sure that this is the coercion that everyone will want? "evaluate the polynomial at 1"?

eric-wieser replied:
I'm reading this as "evaluate my direct sum over A_i back into A" - I hadn't thought of the "evaluate the polynomial" interpretation, and I think it hurts rather than hinders

kbuzzard commented on src/data/finsupp/basic.lean:
```diff
@@ -199,6 +199,16 @@ begin
+  rw [single_apply, single_apply],
+  ext,
+  split_ifs,
+  { rw h, },
```
are those commas OK in mathlib?

bryangingechen replied:
We're not consistent about them, but I think it's fine.

kbuzzard reviewed (commented):
I think that if this is to be the way a grading is defined, it might be good to spell out what the heck is going on in the module docstring -- it took me a while to figure out the relationship between `grades` and what most people think of as a grading (a decomposition of A into a direct sum of A_i).

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into eric-wieser/free_algebra-grading

eric-wieser committed:
wip

eric-wieser committed:
Merge branch 'master' of github.com:leanprover-community/mathlib into eric-wieser/free_algebra-grading

eric-wieser committed:
fix(*): adjust for changed lemmas

eric-wieser committed:
Merge branch 'master' into eric-wieser/free_algebra-grading

eric-wieser commented on src/data/finsupp/basic.lean:
```diff
@@ -200,6 +200,16 @@ begin
   { rw [single_eq_of_ne h, zero_apply] }
 end
 
+lemma single_of_single_apply (a a' : α) (b : M) :
```
Split to #5219
